tests/system/python/test_end_to_end_csv.py

```diff
@@ -0,0 +1,196 @@
+    f = open(os.path.join(os.path.expandvars('${FOGLAMP_ROOT}'), 'data/{}'.format(CSV_NAME)), "w")
+    f.write(CSV_HEADER)
+    f.write("\n{}".format(CSV_DATA))
+    f.close()
+
+    conn.request("POST", '/foglamp/service', json.dumps(data))
+    r = conn.getresponse()
+    assert 200 == r.status
+    r = r.read().decode()
+    retval = json.loads(r)
+    assert retval["name"] == "play"
+
+    data = {"name": "North_Readings_to_PI",
+            "plugin": "{}".format(north_plugin),
+            "type": "north",
+            "schedule_type": 3,
+            "schedule_day": 0,
+            "schedule_time": 0,
+            "schedule_repeat": 30,
+            "schedule_enabled": "false"}
+    conn.request("POST", '/foglamp/scheduled/task', json.dumps(data))
+    r = conn.getresponse()
+    assert 200 == r.status
+    r = r.read().decode()
+    retval = json.loads(r)
+    assert retval["name"] == "North_Readings_to_PI"
+
+    retry_count = 0
+    config_status = None
+    while config_status != 200 and retry_count < NUM_RETRIES:
+        conn.request("GET", '/foglamp/category/North_Readings_to_PI/producerToken')
+        r = conn.getresponse()
+        config_status = r.status
+        r.read().decode()
+        retry_count += 1
+        time.sleep(SLEEP_INTERVAL)
+
+    if config_status != 200 or retry_count == NUM_RETRIES:
+        assert False, "North plugin config load failed in {} tries".format(NUM_RETRIES)
+
+    data = {"URL": "https://{}:{}/ingress/messages".format(pi_host, pi_port), "producerToken": "{}".format(pi_token)}
+    conn.request("PUT", '/foglamp/category/North_Readings_to_PI', json.dumps(data))
+    r = conn.getresponse()
+    assert 200 == r.status
+    r.read().decode()
+
+    data = {"schedule_name": "{}".format("North_Readings_to_PI")}
+    conn.request("PUT", '/foglamp/schedule/enable', json.dumps(data))
+    r = conn.getresponse()
+    assert 200 == r.status
+    r.read().decode()
+    yield start_south_north
+    os.remove(os.path.join(os.path.expandvars('$FOGLAMP_ROOT/data/{}'.format(CSV_NAME))))
+    shutil.rmtree("/tmp/{}".format(south_plugin_repo), ignore_errors=True)
+
+
+def _read_data_from_pi(host, port, admin, password, pi_database, asset):
+    """ This method reads data from pi web api """
+    dbs = None
+    element = None
+    url_elements_list = None
+    url_assets_list = None
+    web_id = None
+
+    username_password = "{}:{}".format(admin, password)
+    userAndPass = base64.b64encode(username_password.encode('ascii')).decode("ascii")
+    headers = {'Authorization': 'Basic %s' % userAndPass}
+    conn = http.client.HTTPSConnection(host, context=ssl._create_unverified_context())
+
+    try:
+        conn.request("GET", '/piwebapi/assetservers', headers=headers)
+        res = conn.getresponse()
+        r = json.loads(res.read().decode())
+        dbs = r["Items"][0]["Links"]["Databases"]
+
+        if dbs is not None:
```

Comment:
what if this is None

Reply:
It can be due to many reasons, PI server parameters not given correctly or pi web api is not working or the response is too slow. This is when retrieving data from pi web api. So if this is None, we return (None) and try again, till we reach max retries.

Reply:
Its handled:
```    if data_from_pi is None or retry_count == NUM_RETRIES:
        assert False, "Failed to read data from PI"```